src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -38,117 +36,111 @@
+            //this constructor is used only by extended classes for smallest unit; upper bound is guarded there accordingly
+
+            validateMinUnit(unit, minUnit, value);
         }
         else
         {
-            throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT.get(minUnit) +
+            throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + acceptedUnits(minUnit) +
                                              " where case matters and only non-negative values are accepted");
         }
     }
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of bytes.
-     *
-     * @param bytes the amount of bytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inBytes(long bytes)
+
+    private DataStorageSpec(String value, DataStorageUnit minUnit, long max)
     {
-        return new DataStorageSpec(bytes, DataStorageUnit.BYTES);
+        this(value, minUnit);
+
+        validateMinUnit(unit, minUnit, value);
+        validateQuantity(value, quantity(), unit(), minUnit, max);
     }
 
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of kibibytes.
-     *
-     * @param kibibytes the amount of kibibytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inKibibytes(long kibibytes)
+    private void validateMinUnit(DataStorageUnit unit, DataStorageUnit minUnit, String value)
     {
-        return new DataStorageSpec(kibibytes, KIBIBYTES);
+        if (unit.compareTo(minUnit) < 0)
+            throw new ConfigurationException(String.format("Invalid data storage: %s Accepted units:%s", value, acceptedUnits(minUnit)));
     }
 
-    /**
-     * Creates a {@code DataStorageSpec} of the specified amount of mebibytes.
-     *
-     * @param mebibytes the amount of mebibytes
-     * @return a {@code DataStorageSpec}
-     */
-    public static DataStorageSpec inMebibytes(long mebibytes)
+    private String acceptedUnits(DataStorageUnit minUnit)
+    {
+        DataStorageUnit[] units = DataStorageUnit.values();
+        return Arrays.toString(Arrays.copyOfRange(units, minUnit.ordinal(), units.length));
+    }
+
+    private static void validateQuantity(String value, long quantity, DataStorageUnit sourceUnit, DataStorageUnit minUnit, long max)
     {
-        return new DataStorageSpec(mebibytes, MEBIBYTES);
+        // no need to validate for negatives as they are not allowed at first place from the regex
+
+        if (minUnit.convert(quantity, sourceUnit) >= max)
+            throw new ConfigurationException("Invalid data storage: " + value + ". It shouldn't be more than " +
+                                             (max - 1) + " in " + minUnit.name().toLowerCase());
     }
 
+    private static void validateQuantity(long quantity, DataStorageUnit unit, DataStorageUnit minUnit, long max)
```

Comment:
Nit: I'd rename the `unit` argument to `sourceUnit` to keep consistency with other methods:
```suggestion
    private static void validateQuantity(long quantity, DataStorageUnit sourceUnit, DataStorageUnit minUnit, long max)
```

Reply:
Sure, changed also in `validateMinUnit`